src/main/scala/org/apache/spark/sql/execution/datasources/greenplum/GreenplumUtils.scala

```diff
@@ -60,71 +61,166 @@ object GreenplumUtils extends Logging {
+    }
+    val accumulator = finalDf.sparkSession.sparkContext.longAccumulator("copySuccess")
+    val partNum = finalDf.rdd.getNumPartitions
 
-    def convertRow(row: Row): Array[Byte] = {
-      var i = 0
-      val values = new Array[String](schema.length)
-      while (i < schema.length) {
-        if (!row.isNullAt(i)) {
-          values(i) = convertValue(valueConverters(i).apply(row, i))
-        } else {
-          values(i) = "NULL"
-        }
-        i += 1
-      }
-      (values.mkString(options.delimiter) + "\n").getBytes("UTF-8")
+    finalDf.foreachPartition { rows =>
+      copyParition(rows, options, schema, options.table, accumulator)
+    }
+
+    if (accumulator.value != partNum) {
+      throw new PartitionCopyFailureException(
+        s"Some partitions failed(successful: ${accumulator.value}, total: $partNum)")
     }
+  }
+
+  /**
+   * https://www.postgresql.org/docs/9.2/sql-copy.html
+   *
+   * Copy data to greenplum and overwrite relative gptable,
+   * which can be dropped or does not exist.
+   *
+   * @param df the [[DataFrame]] will be copy to the Greenplum
+   * @param schema the table schema in Greemnplum
+   * @param options Options for the Greenplum data source
+   */
+  def copyOverwriteToGreenplum(
+      df: DataFrame,
+      schema: StructType,
+      options: GreenplumOptions): Unit = {
+    val randomString = UUID.randomUUID().toString.filterNot(_ == '-')
+    val suffix = "sparkGpTmp"
+    val tempTable = s"${options.table}_${randomString}_$suffix"
+    val strSchema = JdbcUtils.schemaString(df, options.url, options.createTableColumnTypes)
+    val createTempTbl = s"CREATE TABLE $tempTable ($strSchema) ${options.createTableOptions}"
 
-    def convertValue(str: String): String = {
-      assert(options.delimiter.length == 1, "The delimiter should be a single character.")
-      val delimiter = options.delimiter.charAt(0)
-      str.flatMap {
-        case '\\' => "\\\\"
-        case '\n' => "\\n"
-        case '\r' => "\\r"
-        case `delimiter` => s"\\$delimiter"
-        case c => s"$c"
+    val conn = JdbcUtils.createConnectionFactory(options)()
+
+    try {
+      executeStatement(conn, createTempTbl)
+      val accumulator = df.sparkSession.sparkContext.longAccumulator("copySuccess")
+      val partNum = df.rdd.getNumPartitions
+
+      df.foreachPartition { rows =>
+        copyParition(rows, options, schema, tempTable, accumulator)
       }
+
+      if (accumulator.value == partNum) {
+        val dropTbl = s"DROP TABLE IF EXISTS ${options.table}"
+        executeStatement(conn, dropTbl)
+
+        val renameTempTbl = s"ALTER TABLE $tempTable RENAME TO ${options.table}"
+        executeStatement(conn, renameTempTbl)
+      } else {
+        val dropTempTbl = s"DROP TABLE $tempTable"
+        executeStatement(conn, dropTempTbl)
+        throw new PartitionCopyFailureException(
+          s"Some partitions failed(successful: ${accumulator.value}, total: $partNum)")
```

Comment:
we shall notify users that the whole job has been aborted  here

Reply:
I have add a exception type `PartitionCopyFailureException`, and throw a  PartitionCopyFailureException instance when we find out  there are failed partitions.

Reply:
I mean the error message can be a little improved, may be tell users that "Job aborted ", "Retry again " etc..